lib/job.js

```diff
@@ -1,106 +1,249 @@
-  failure: 'failure',
-  dependencies: 'dependencies',
-  dependents: 'dependents',
-  spawnedFromJid: 'spawned_from_jid',
-  expiresAt: 'expires',
-  klassName: 'klass',
-  queueName: 'queue',
-  workerName: 'worker',
-  originalRetries: 'retries',
-  retriesLeft: 'remaining',
-  rawQueueHistory: 'history',
-};
-
-
-// TODO: split up into BaseJob and RecurringJob
-class Job {
-  constructor(client, args) {
-    this.client = client;
-    this._stateChanged = false;
-
-    // Copy args
-    Object.assign(this, _.mapValues(JOB_ARGS_TO_PROPERTIES, key => args[key]));
-
-    // Some properties need extra parsing / fixing
-    this.data = JSON.parse(this.data);
-    // Side effect of Lua doing JSON parsing -- Lua empty array is {}
-    if (_.isEmpty(this.tags)) this.tags = [];
-    if (_.isEmpty(this.dependents)) this.dependents = [];
-    if (_.isEmpty(this.dependencies)) this.dependencies = [];
-  }
+      // If there's no such attribute, break out of the loop and throw
+      if (!result) {
+        break;
+      }
 
-  // If the job has been fail()ed, complete()d, etc. since object creation
-  isStateChanged() {
-    return this._stateChanged;
+      logger.debug('Found "%s"', result.name);
+      return result;
+    } catch (err) {
+      attributes.push(segments.pop());
+    }
   }
+  throw Error('Could not find ' + name);
+}
 
-  getKlass() {
-    return klassFinder.findClass(this.klassName);
-  }
+Job.prototype.setPriority = function(priority) {
+  let self = this;
+  return this.client.call('priority', this.jid, priority)
+    .then(() => {
+      self.priority = priority;
+    })
+}
 
-  // Equivalent to qless-ruby's Job#perform -- runs job
-  // and passes any errors from job up to called.
-  // cb is called with error if the job fails. (Note that
-  // this is different from qless-py's Job#process)
-  // TODO: around_perform
-  perform(cb) {
-    const klass = this.getKlass();
-    if (!klass) return cb(new errors.CouldntLoadClass(this.klassName));
-    if (typeof(klass.perform) !== 'function') return cb(new errors.ClassLackingPerformMethod(this.klassName));
-    klass.perform(this, cb);
-  }
+Job.prototype.getQueue = function() {
+  return this.client.queue(this.queueName);
+}
 
-  toString() {
-    return `<qless.Job ${this.klassName} (${this.jid} / ${this.queueName} / ${this.state} / ${JSON.stringify(this.data)})>`;
+/**
+ * Get the class that corresponds to this klassName.
+ *
+ * A note on security -- abolute paths or those starting with a . are not
+ * allowed. By disallowing them, we require that the job must be available in
+ * node_modules, which allows for some amount of sandboxing. This, of course,
+ * does not preclude the possibility of job code importing arbitrary modules.
+ * It merely means that only job modules that have been installed in this
+ * environment are available.
+ */
+Job.prototype.getKlass = function() {
+  let name = Path.normalize(this.klassName);
+  if (name.startsWith('.') || name.startsWith('/')) {
```

Comment:
Just to close the loop here, too, because of the `Path.normalize`, we can check for an indication of being an absolute path (starting with `/`) or a relative path (starting with `.`). Everything else would be interpreted as relative to `node_modules`.